src/main/java/org/spongepowered/common/service/scheduler/AsyncScheduler.java

```diff
@@ -194,520 +113,73 @@ private void recalibrateMinimumTimeout() {
-                // start it after the interval (period) of the repeating task parameter.
-                if (task.state == ScheduledTask.ScheduledTaskState.WAITING) {
-                    threshold = task.offset;
-                } else if (task.state == ScheduledTask.ScheduledTaskState.RUNNING) {
-                    threshold = task.period;
-                }
-
-                // This moment is 'now'
-                long now = System.currentTimeMillis();
-
-                // So, if the current time minus the timestamp of the task is greater than
-                // the delay to wait before starting the task, then start the task.
-
-                // Repeating tasks get a reset-timestamp each time they are set RUNNING
-                // If the task has a period of 0 (zero) this task will not repeat, and is removed
-                // after we start it.
-
-                if (threshold <= (now - task.timestamp)) {
-                    // startTask is just a utility function within the Scheduler that
-                    // starts the task.
-                    // If the task is not a one time shot then keep it and
-                    // change the timestamp to now.  It is a little more
-                    // efficient to do this now than after starting the task.
-                    task.timestamp = System.currentTimeMillis();
-                    boolean bTaskStarted = startTask(task);
-                    if (bTaskStarted) {
-                        task.setState(ScheduledTask.ScheduledTaskState.RUNNING);
-                        // If task is one time shot, remove it from the list.
-                        if (task.period == 0L) {
-                            this.taskMap.remove(task.getUniqueId());
-                        }
-                    }
-                }
-            }
-            this.lastProcessingTimestamp = System.currentTimeMillis();
-        } finally {
-            this.lock.unlock();
-        }
-    }
-
-    private Optional<Task> utilityForAddingAsyncTask(ScheduledTask task) {
-        Optional<Task> resultTask = Optional.absent();
-
-        task.setTimestamp(System.currentTimeMillis());
+    @Override
+    protected void preTick() {
         this.lock.lock();
         try {
-            this.taskMap.put(task.getUniqueId(), task);
-            this.condition.signalAll();
-            resultTask = Optional.of((Task) task);
-        } finally {
-            this.lock.unlock();
+            this.condition.await(this.minimumTimeout, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            // The taskMap has been modified; there is work to do.
+            // Continue on without handling the Exception.
+        } catch (IllegalMonitorStateException e) {
+            Sponge.getLogger().error("The scheduler internal state machine suffered a catastrophic error.  Check #spongedev");
+            Sponge.getLogger().error(e.toString());
```

Comment:
Make sure to give the exception to the logger above, then you can leave out this line. Also a better error message of what exactly happened would be good